lib/compress/zstd_lazy.c

```diff
@@ -845,11 +851,668 @@ FORCE_INLINE_TEMPLATE size_t ZSTD_HcFindBestMatch_extDict_selectMLS (
+    const U32 withinMaxDistance = (curr - lowestValid > maxDistance) ? curr - maxDistance : lowestValid;
+    const U32 isDictionary = (ms->loadedDictEnd != 0);
+    const U32 lowLimit = isDictionary ? lowestValid : withinMaxDistance;
+    U32 nbAttempts = 1U << cParams->searchLog;
+    size_t ml=4-1;
+
+    if (dictMode == ZSTD_dedicatedDictSearch) {
+        const ZSTD_matchState_t* const dms = ms->dictMatchState;
+        const U32 ddsHashLog = dictMode == ZSTD_dedicatedDictSearch
+                         ? dms->cParams.hashLog - ZSTD_LAZY_DDSS_BUCKET_LOG : 0;
+        const size_t ddsIdx = dictMode == ZSTD_dedicatedDictSearch
+                        ? ZSTD_hashPtr(ip, ddsHashLog, mls) << ZSTD_LAZY_DDSS_BUCKET_LOG : 0;
+        const U32* entry = &dms->hashTable[ddsIdx];
+        PREFETCH_L1(entry);
+    }
+
+    /* Update the hashTable and tagTable up to (but not including) ip */
+    ZSTD_row_update_internal(ms, ip, mls, rowLog, rowMask, 1 /* useCache */);
+    {
+        /* Get the hash for ip, compute the appropriate row */
+        U32 const hash = ZSTD_row_nextCachedHash(hashCache, hashTable, tagTable, base, curr, hashLog, rowLog, mls);
+        U32 const relRow = (hash >> kHashTagBits) << rowLog;
+        U32 const tag = hash & kHashTagMask;
+        U32* const row = hashTable + relRow;
+        BYTE* tagRow = (BYTE*)(tagTable + relRow);
+        U32 const head = *tagRow & rowMask;
+        U32 matchBuffer[kRowEntries32];
+        size_t numMatches = 0;
+        size_t currMatch = 0;
+        ZSTD_VecMask matches = 0;
+
+        /* Generate a "matches" bitfield. The nth bit == 1 if the newly-computed "tag" matches the hash at the nth
+           position in a row of the tagTable */
+#ifdef __SSE2__
+        if (rowEntries == 16) {
+            ZSTD_Vec128 hashes = ZSTD_Vec128_read(tagRow + kHashOffset);
+            ZSTD_Vec128 hash1  = ZSTD_Vec128_set8((BYTE)tag);
+            ZSTD_Vec128 cmpeq  = ZSTD_Vec128_cmp8(hashes, hash1);
+            matches            = ZSTD_Vec128_mask8(cmpeq);
+        } else if (rowEntries == 32) {
+            ZSTD_Vec256 hashes = ZSTD_Vec256_read(tagRow + kHashOffset);
+            ZSTD_Vec256 hash1  = ZSTD_Vec256_set8((BYTE)tag);
+            ZSTD_Vec256 cmpeq  = ZSTD_Vec256_cmp8(hashes, hash1);
+            matches            = ZSTD_Vec256_mask8(cmpeq);
+        } else {
+            assert(0);
+        }
+#else
+        if (rowEntries == 16) {
+            ZSTD_Vec128 hashes = {0};
+            ZSTD_Vec128 vec = {0};
+            ZSTD_Vec128_scalarRead(tagRow + kHashOffset, hashes);
+            ZSTD_Vec128_scalarSet(tag, vec);
+            matches  = ZSTD_Vec128_scalarCmp(hashes, vec);
+        } else if (rowEntries == 32) {
+            ZSTD_Vec256 hashes = {0};
+            ZSTD_Vec256 vec = {0};
+            ZSTD_Vec256_scalarRead(tagRow + kHashOffset, hashes);
+            ZSTD_Vec256_scalarSet(tag, vec);
+            matches  = ZSTD_Vec256_scalarCmp(hashes, vec);
+        }
```

Comment:
Since we're doing exactly the same operations, would it make more sense to keep the functions the same name & signature, so we don't need an `ifdef __SSE2__` here?

You may need to combine `cmp8()` and `mask8()` into a single `cmpMask8()` function. And you may need to change `ZSTD_Vec128` from an array to a struct containing an array to make it work. But, otherwise the compiler should be able to figure it all out, since everything is inlined.